src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,329 @@
+                        MinUpperBound = b.Time;
+                    }
+                }
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// QubitTimeNodes sorted by Time field. SortedSet won't accept values with the same Time field.
+        /// When QubitPool needs to store multiple qubits with the same Time, additional QubitTimeNode instances
+        /// are linked to the head node from this set.
+        /// </summary>
+        private readonly SortedSet<QubitTimeNode> QubitsSortedByTime;
+
+        /// <summary>
+        /// Comparer of two nodes, which compares them by Time field and collects additional data.
+        /// </summary>
+        private readonly VisitingComparer NodeComparer;
+
+        /// <summary>
+        /// Sample which is been compared to the nodes stored in the set.
+        /// Multiple comparisons are done by SortedSet to find if this Sample is present in the set.
+        /// </summary>
+        private readonly QubitTimeNode Sample = new QubitTimeNode(); // We reuse same object to avoid allocations.
+
+        public SortedQubitPool() {
+            NodeComparer = new VisitingComparer();
+            QubitsSortedByTime = new SortedSet<QubitTimeNode>(NodeComparer);
+        }
+
+        /// <summary>
+        /// Add Qubit to the set with specified time. Multiple qubits with the same time can be added.
+        /// Complexity is log(N), where N - number of nodes.
+        /// </summary>
+        /// <param name="qubitId">Id of a qubit to add to the set.</param>
+        /// <param name="time">Time of the qubit.</param>
+        public void Add(long qubitId, ComplexTime time) {
+            QubitTimeNode newNode = new QubitTimeNode(qubitId, time);
+            if (QubitsSortedByTime.Add(newNode)) {
+                // New item was added to the set, we are done.
+                return;
+            }
+            if (!QubitsSortedByTime.TryGetValue(newNode, out QubitTimeNode existingNode)) {
+                // We cannot add, but we cannot find a node with the same value. Is this a floating point glitch?
+                Debug.Assert(false, "Cannot add a value to SortedSet<QubitTimeNode> that isn't in the set.");
+                return;
+            }
+            // Add new node to the linked list as a second element.
+            newNode.NextNode = existingNode.NextNode;
+            existingNode.NextNode = newNode;
+        }
+
+        /// <summary>
+        /// Find qubit with the specified time in the set. If such qubit is not present,
+        /// find either maximum value less than it or minimum value greater than it.
+        /// This function uses one call to SortedSet.TryGetValue() so it takes log(N) time, where N - number of nodes.
+        /// </summary>
+        /// <param name="requestedTime">Sample time to find</param>
+        /// <param name="getLowerBound">"true" to find maximum value less than sample</param>
```

Comment:
As stated, is sounds like an opt-in: "if you want a lower bound in case the exact match cannot be found, pass in `true`", but in reality it seems to mean: "if cannot find the exact match we'll return either lower or upper bound, choose which of the two you prefer"... Maybe, name it `preferLowerBound`?

Reply:
I updated the comment to indicate what both "true" and "false" means.